compiler/hash-ast/src/visitor.rs

```diff
@@ -0,0 +1,1579 @@
+                .collect(),
+            entries: node
+                .entries
+                .iter()
+                .map(|e| visitor.visit_struct_literal_entry(e.ast_ref()))
+                .collect(),
+        }
+    }
+
+    pub struct StructLiteralEntry<'c, V: AstVisitor<'c>> {
+        pub name: V::NameRet,
+        pub value: V::ExpressionRet,
+    }
+
+    pub fn walk_struct_literal_entry<'c, V: AstVisitor<'c>>(
+        visitor: &mut V,
+        node: ast::AstNodeRef<ast::StructLiteralEntry<'c>>,
+    ) -> StructLiteralEntry<'c, V> {
+        StructLiteralEntry {
+            name: visitor.visit_name(node.name.ast_ref()),
+            value: visitor.visit_expression(node.value.ast_ref()),
+        }
+    }
+
+    pub enum Expression<'c, V: AstVisitor<'c>> {
+        FunctionCall(V::FunctionCallExprRet),
+        Intrinsic(V::IntrinsicKeyRet),
+        Variable(V::VariableExprRet),
+        PropertyAccess(V::PropertyAccessExprRet),
+        Ref(V::RefExprRet),
+        Deref(V::DerefExprRet),
+        LiteralExpr(V::LiteralExprRet),
+        Typed(V::TypedExprRet),
+        Block(V::BlockExprRet),
+        Import(V::ImportExprRet),
+    }
+
+    pub fn walk_expression<'c, V: AstVisitor<'c>>(
+        visitor: &mut V,
+        node: ast::AstNodeRef<ast::Expression<'c>>,
+    ) -> Expression<'c, V> {
+        match node.kind() {
+            ast::ExpressionKind::FunctionCall(inner) => {
+                Expression::FunctionCall(visitor.visit_function_call_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Intrinsic(inner) => {
+                Expression::Intrinsic(visitor.visit_intrinsic_key(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Variable(inner) => {
+                Expression::Variable(visitor.visit_variable_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::PropertyAccess(inner) => Expression::PropertyAccess({
+                visitor.visit_property_access_expr(node.with_body(inner))
+            }),
+            ast::ExpressionKind::Ref(inner) => {
+                Expression::Ref(visitor.visit_ref_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Deref(inner) => {
+                Expression::Deref(visitor.visit_deref_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::LiteralExpr(inner) => {
+                Expression::LiteralExpr(visitor.visit_literal_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Typed(inner) => {
+                Expression::Typed(visitor.visit_typed_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Block(inner) => {
+                Expression::Block(visitor.visit_block_expr(node.with_body(inner)))
+            }
+            ast::ExpressionKind::Import(inner) => {
+                Expression::Import(visitor.visit_import_expr(node.with_body(inner)))
+            }
+        }
```

Comment:
It would be nice if u could do something like:
```rs
ExpressionKind:{variant}(inner) =>
        Expression::{variant}(...)
```

Reply:
Yeah, that would be a macro.